File: app/src/free/java/co/smartreceipts/android/ad/region/RegionChecker.java
package co.smartreceipts.android.ad.region;

import android.content.Context;
import android.support.annotation.NonNull;
import android.telephony.TelephonyManager;

import com.google.common.base.Preconditions;

import java.util.Arrays;
import java.util.List;
import java.util.Locale;

import javax.inject.Inject;

import co.smartreceipts.android.di.scopes.ApplicationScope;

@ApplicationScope
public class RegionChecker {

    private static final List<String> EU_ISO_3166_COUNTRIES = Arrays.asList("AT", "BE", "BG", "CY", "CZ", "DE", "DK", "EE", "ES", "FI", "FR", "GB", "GR", "HR", "HU", "IE", "IT", "LT", "LU", "LV", "MT", "NL", "PL", "PT", "RO", "SE", "SI", "SK", "UK");

    private final Context context;

    @Inject
    public RegionChecker(@NonNull Context context) {
        this.context = Preconditions.checkNotNull(context);
    }

    public boolean isInWesternEurope() {
        final String iso3166CountryCode = getUserCountry(context);
        return iso3166CountryCode != null && EU_ISO_3166_COUNTRIES.contains(iso3166CountryCode);
    }

    /**
     * Get ISO 3166-1 alpha-2 country code for this device (or {@code null} if not available)
     *
     * @param context Context reference to get the TelephonyManager instance from
     * @return country code or {@code null}
     */
    private String getUserCountry(@NonNull Context context) {
        try {
            final TelephonyManager tm = (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);
            final String simCountry = tm.getSimCountryIso();
            if (simCountry != null && simCountry.length() == 2) { // SIM country code is available
                return simCountry.toLowerCase(Locale.US);
            }
            else if (tm.getPhoneType() != TelephonyManager.PHONE_TYPE_CDMA) { // device is not 3G (would be unreliable)
                String networkCountry = tm.getNetworkCountryIso();
                if (networkCountry != null && networkCountry.length() == 2) { // network country code is available
                    return networkCountry.toLowerCase(Locale.US);
                }
            }
        }
        catch (Exception e) {
            // Intentional no-op to avoid crash
        }
        return null;
    }


}

File: app/src/main/java/co/smartreceipts/android/tooltip/privacy/RegionChecker.kt
package co.smartreceipts.android.tooltip.privacy

import android.content.Context
import android.telephony.TelephonyManager
import co.smartreceipts.core.di.scopes.ApplicationScope
import java.util.*
import javax.inject.Inject


/**
 * Allows us to quickly check a specific region for scenarios in which app business logic differs as
 * per the user's region
 */
@ApplicationScope
class RegionChecker internal constructor(private val telephonyManager: TelephonyManager?) {

    @Inject
    constructor(context: Context) : this(context.getSystemService(Context.TELEPHONY_SERVICE) as? TelephonyManager)

    /**
     * Checks if the user's phone is an EU model (useful for GDPR prompts)
     *
     * @return true if this is an EU model phone. false otherwise
     */
    fun isInTheEuropeanUnion() : Boolean {
        return EU_COUNTRY_CODES.contains(getUserIso3166CountryCode())
    }

    /**
     * Gets the two-character country iso code for the current user based on a combination of their
     * telephone network, SIM card, and default locale
     *
     * @return the Iso 3166 country country code
     */
    private fun getUserIso3166CountryCode() : String {
        // First, check the country code from the mobile network
        // Note: The docs state that this may be unreliable on CDMA networks, so we ignore this check here
        if (telephonyManager?.phoneType != TelephonyManager.PHONE_TYPE_CDMA) {
            val networkCountryIso = telephonyManager?.networkCountryIso
            if (networkCountryIso?.isNotEmpty() == true) {
                return networkCountryIso.toUpperCase(Locale.US)
            }
        }

        // Next, check it from the SIM card
        val simCountryIso = telephonyManager?.simCountryIso
        if (simCountryIso?.isNotEmpty() == true) {
            return simCountryIso.toUpperCase(Locale.US)
        }

        // Finally, return the current locale
        return Locale.getDefault().country.toUpperCase(Locale.US)
    }

    companion object {
        /**
         * This array contains a list of all known EU country codes
         *
         * From: https://ec.europa.eu/eurostat/statistics-explained/index.php/Tutorial:Country_codes_and_protocol_order
         */
        private val EU_COUNTRY_CODES = arrayListOf(
                "BE",
                "BG",
                "CZ",
                "DK",
                "DE",
                "EE",
                "IE",
                "EL",
                "ES",
                "FR",
                "HR",
                "IT",
                "CY",
                "LV",
                "LT",
                "LU",
                "HU",
                "MT",
                "NL",
                "AT",
                "PL",
                "PT",
                "RO",
                "SI",
                "SK",
                "FI",
                "SE",
                "UK",
                "IS",
                "LI",
                "NO",
                "CH"
        )
    }
}